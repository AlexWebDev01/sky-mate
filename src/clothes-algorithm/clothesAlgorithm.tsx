import { TEMPERATURE_RANGES, CLOTHES_ADVICES } from "./clothesAlgorithmData";

const getTempRange = (temp: number): string => {
  for (let i = 0; i < TEMPERATURE_RANGES.length; i++) {
    if (temp < TEMPERATURE_RANGES[i].highestTemp) {
      return TEMPERATURE_RANGES[i].name;
    }
  }

  return "";
};

export const calculateClothesAdvice = (
  weatherCondition: string,
  temp: number
) => {
  const clothesAdvices = CLOTHES_ADVICES[weatherCondition];
  const tempRangeName = getTempRange(temp);

  if (clothesAdvices[tempRangeName]) {
    return { adviceData: clothesAdvices[tempRangeName], tempRangeName };
  } else {
    console.log(`For now there is no such conditions. 
        Temperature: ${temp}, 
        Temperature range: ${tempRangeName}, 
        Weather condition: ${weatherCondition}
        `);

    return null;
  }
};
