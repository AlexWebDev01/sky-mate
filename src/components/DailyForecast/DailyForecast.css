@import url("../../shared/assets/fonts/fonts.css");
.daily-forecast {
  display: grid;
  grid-template-columns: 1fr 1fr 1fr 1fr;
  grid-template-rows: 1fr 1fr;
  align-items: center;
  width: 455px;
  height: 224px;
  border-radius: 40px;
  background: #fff;
  box-shadow: 0px 4px 30px 0px rgba(0, 0, 0, 0.14);
  padding: 47px 59px 52px 59px;
  font-family: Oddval;
  box-sizing: border-box;
  z-index: 1;
  transition: width 0.5s ease, height 0.5s ease;
  overflow: hidden;
}
.daily-forecast .daily-weather {
  display: flex;
  flex-direction: column;
  align-items: center;
  width: 60px;
  height: 50px;
  border-radius: 14px;
  padding: 7px 12px;
  box-sizing: content-box;
  text-align: left;
}
.daily-forecast .daily-weather .day {
  color: #000;
  font-size: 24px;
  font-weight: 600;
}
.daily-forecast .daily-weather .day-temperature {
  color: #6e6e6e;
  font-size: 20px;
  font-weight: 600;
  line-height: 0.6rem;
}
.daily-forecast .daily-weather .date,
.daily-forecast .daily-weather .night-temperature,
.daily-forecast .daily-weather .weather {
  display: none;
}

.forecast-expanded {
  display: grid;
  grid-template-columns: 1fr 1fr 1fr 1fr;
  grid-template-rows: 1fr 1fr;
  align-items: center;
  width: 455px;
  height: 468px;
  border-radius: 40px;
  background: #fff;
  box-shadow: 0px 4px 30px 0px rgba(0, 0, 0, 0.14);
  padding: 53px 50px;
  font-family: Oddval;
  box-sizing: border-box;
  z-index: 1;
  transition: width 0.5s ease, height 0.5s ease;
  overflow: hidden;
}
.forecast-expanded .daily-weather {
  display: flex;
  flex-direction: column;
  width: 85px;
  height: 165px;
  border-radius: 14px;
  padding: 7px 12px;
  box-sizing: border-box;
  text-align: left;
}
.forecast-expanded .daily-weather .day {
  color: #000;
  font-size: 24px;
  font-weight: 600;
}
.forecast-expanded .daily-weather .date {
  font-size: 16px;
  font-weight: 600;
  text-transform: lowercase;
  margin-bottom: 22px;
}
.forecast-expanded .daily-weather .day-temperature {
  font-size: 20px;
  font-weight: 600;
}
.forecast-expanded .daily-weather .night-temperature {
  color: #d9d9d9;
  font-size: 20px;
  font-weight: 600;
}
.forecast-expanded .daily-weather .weather {
  color: #000;
  font-size: 16px;
  font-weight: 600;
}
.forecast-expanded .rain,
.forecast-expanded .snow {
  color: var(--blue);
}
.forecast-expanded .sun,
.forecast-expanded .clear {
  color: var(--orange);
}
.forecast-expanded .clouds,
.forecast-expanded .fog,
.forecast-expanded .dust,
.forecast-expanded .mist {
  color: var(--green);
}

.daily-forecast {
  cursor: pointer;
}

.forecast-expanded:hover {
  cursor: default;
}/*# sourceMappingURL=DailyForecast.css.map */