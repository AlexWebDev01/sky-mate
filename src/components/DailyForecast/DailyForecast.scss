@import url("../../shared/assets/fonts/fonts.css");

.daily-forecast {
  display: grid;
  grid-template-columns: 1fr 1fr 1fr 1fr;
  grid-template-rows: 1fr 1fr;
  align-items: center;
  width: 455px;
  height: 224px;
  border-radius: 40px;
  background: #fff;
  box-shadow: 0px 4px 30px 0px rgba(0, 0, 0, 0.14);
  padding: 47px 59px 52px 59px;
  font-family: Oddval;
  box-sizing: border-box;
  z-index: 1;
  transition: width 0.5s ease, height 0.5s ease;
  overflow: hidden;

  .daily-weather {
    display: flex;
    flex-direction: column;
    align-items: center;
    width: 60px;
    height: 50px;
    border-radius: 14px;
    padding: 7px 12px;
    box-sizing: content-box;
    text-align: left;

    .day {
      color: #000;
      font-size: 24px;
      font-weight: 600;
    }

    .day-temperature {
      color: #6e6e6e;
      font-size: 20px;
      font-weight: 600;
      line-height: 0.6rem;
    }

    .date,
    .night-temperature,
    .weather {
      display: none;
    }
  }
}

.forecast-expanded {
  display: grid;
  grid-template-columns: 1fr 1fr 1fr 1fr;
  grid-template-rows: 1fr 1fr;
  align-items: center;
  width: 455px;
  height: 468px;
  border-radius: 40px;
  background: #fff;
  box-shadow: 0px 4px 30px 0px rgba(0, 0, 0, 0.14);
  padding: 53px 50px;
  font-family: Oddval;
  box-sizing: border-box;
  z-index: 1;
  transition: width 0.5s ease, height 0.5s ease;
  overflow: hidden;

  .daily-weather {
    display: flex;
    flex-direction: column;
    width: 85px;
    height: 165px;
    border-radius: 14px;
    padding: 7px 12px;
    box-sizing: border-box;
    text-align: left;

    .day {
      color: #000;
      font-size: 24px;
      font-weight: 600;
    }

    .date {
      font-size: 16px;
      font-weight: 600;
      text-transform: lowercase;
      margin-bottom: 22px;
    }

    .day-temperature {
      font-size: 20px;
      font-weight: 600;
    }

    .night-temperature {
      color: #d9d9d9;
      font-size: 20px;
      font-weight: 600;
    }

    .weather {
      color: #000;
      font-size: 16px;
      font-weight: 600;
    }
  }

  .rain,
  .snow {
    color: #67a8f4;
  }

  .sun,
  .clear {
    color: #ff753e;
  }

  .clouds,
  .fog,
  .dust,
  .mist {
    color: #29af77;
  }
}

.daily-forecast {
  cursor: pointer;
}

.forecast-expanded:hover {
  cursor: default;
}
