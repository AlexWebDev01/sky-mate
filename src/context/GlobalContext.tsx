import { useState, createContext, useContext, FunctionComponent } from "react";
import {
  Coordinates,
  GlobalContext,
  GlobalProviderProps,
} from "./GlobalContext.interface";

const GlobalContext = createContext<GlobalContext | null>(null);

export const useGlobalContext = () => {
  const context = useContext(GlobalContext);

  if (!context) {
    throw new Error("useWords must be used within a WordsProvider");
  }

  return context;
};

export const GlobalProvider: FunctionComponent<GlobalProviderProps> = ({
  children,
}) => {
  const [coordinates, setCoordinates] = useState<Coordinates | null>(null);
  const [weatherData, setWeatherData] = useState<object | null>(null);
  const [location, setLocation] = useState<string | null>(null);
  const [pageStyle, setPageStyle] = useState<string>("");

  const sharedState: GlobalContext = {
    coordinates,
    setCoordinates,
    weatherData,
    setWeatherData,
    location,
    setLocation,
    pageStyle,
    setPageStyle,
  };

  return (
    <GlobalContext.Provider value={sharedState}>
      {children}
    </GlobalContext.Provider>
  );
};
