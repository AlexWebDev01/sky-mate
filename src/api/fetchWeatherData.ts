import { Coordinates } from "../context/GlobalContext.interface";

export const fetchWeatherData = async (
  coordinates: Coordinates,
  units: string
) => {
  console.log("COORDINATES:", coordinates);

  const response = await fetch(
    `${import.meta.env.VITE_WEATHER_APP_API_URL}/onecall?lat=${
      coordinates.lat
    }&lon=${coordinates.lon}&units=${units}&exclude=minutely&appid=${
      import.meta.env.VITE_WEATHER_APP_API_KEY
    }`
  );

  const data = await response.json();

  return data;
};
