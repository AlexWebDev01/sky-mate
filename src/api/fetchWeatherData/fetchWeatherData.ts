import { Units, Coordinates, WeatherData } from "./fetchWeatherData.interface";

export const fetchWeatherData = async (
  coordinates: Coordinates,
  units: Units
): Promise<WeatherData> => {
  const response = await fetch(
    `${import.meta.env.VITE_WEATHER_APP_API_URL}/onecall?lat=${
      coordinates.lat
    }&lon=${coordinates.lon}&units=${units}&exclude=minutely&appid=${
      import.meta.env.VITE_WEATHER_APP_API_KEY
    }`
  );

  const data: WeatherData = await response.json();

  return data;
};
