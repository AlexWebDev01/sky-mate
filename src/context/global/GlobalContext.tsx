import {
  useState,
  createContext,
  useContext,
  FunctionComponent,
  useCallback,
  useMemo,
} from 'react';
import {
  GlobalContext as GlobalContextInterface,
  GlobalProviderProps,
  GlobalState,
} from './GlobalContext.interface';

import { fetchLocationData } from '@api/fetchLocationData/fetchLocationData';
import { fetchUserLocatioByIP } from '@api/fetchUserLocationByIP/fetchUserLocationByIP';
import { separateCoordinates } from '@shared/helpers/date/date';
import { fetchWeatherData } from '@api/fetchWeatherData/fetchWeatherData';
import { calculateClothesAdvice } from '@shared/helpers/clothesAlgorithm';

import { WeatherConditions } from '@shared/constants/clothesAlgorithm/clothesAlgorithm.interface';
import { LocationDataByIP } from '@api/fetchUserLocationByIP/fetchUserLocationByIP.interface';
import { LocationData } from '@api/fetchLocationData/fetchLocationData.interface';
import {
  Units,
  Coordinates,
} from '@api/fetchWeatherData/fetchWeatherData.interface';

const GlobalContext = createContext<GlobalContextInterface | null>(null);

export const useGlobalContext = () => {
  const context = useContext(GlobalContext);

  if (!context) {
    throw new Error('Global context should be used in GlobalProvider!');
  }

  return context;
};

export const GlobalProvider: FunctionComponent<GlobalProviderProps> = ({
  children,
}) => {
  const [state, setState] = useState<GlobalState>({
    isLoading: true,
    coordinates: null,
    location: null,
    weatherData: null,
    pageStyle: '',
    clothesAdvice: null,
  });
  const [error, setError] = useState<string>('');

  const getWeatherData = useCallback(async (coordinates: Coordinates) => {
    const data = await fetchWeatherData(coordinates, Units.metric);
    const weatherCondition =
      data.daily[0].weather[0].main.toLowerCase() as WeatherConditions;
    const averageTemp = data.daily[0].temp.day;
    const clothesAdvice = calculateClothesAdvice(weatherCondition, averageTemp);
    return {
      weatherData: data,
      clothesAdvice,
      pageStyle: weatherCondition,
    };
  }, []);

  const fetchData = useCallback(async () => {
    console.group('FETCH DATA');
    console.info('Started initial data fetching');

    setState((prevState) => ({ ...prevState, isLoading: true }));

    try {
      const locationData: LocationDataByIP = await fetchUserLocatioByIP();
      const coordinates = separateCoordinates(locationData.loc);

      const { weatherData, clothesAdvice, pageStyle } =
        await getWeatherData(coordinates);
      setState((prevState) => ({
        ...prevState,
        weatherData,
        clothesAdvice,
        pageStyle,
        coordinates,
        location: locationData.city,
        isLoading: false,
      }));

      console.info('Fetched & calculated data: ', {
        weatherData,
        clothesAdvice,
        pageStyle,
        city: locationData.city,
      });
      console.info('Finished initial data fetching');
      console.groupEnd();
    } catch (error) {
      if (error instanceof Error) {
        setError('Oops, unexpected error!');
      }

      console.error('Failed to fetch initial data:', error);
    }
  }, [getWeatherData]);

  const handleSearch = useCallback(
    async (location: string) => {
      try {
        console.group('HANDLE SEARCH');
        console.info('Started fetching data for location:', location);
        const locationData: LocationData = await fetchLocationData(location);
        const { name, lat, lon } = locationData;

        const { weatherData, clothesAdvice, pageStyle } = await getWeatherData({
          lat,
          lon,
        });

        setState((prevState) => ({
          ...prevState,
          coordinates: { lat, lon },
          location: name,
          weatherData,
          clothesAdvice,
          pageStyle,
        }));
        console.info('Fetched & calculated data: ', {
          weatherData,
          clothesAdvice,
          pageStyle,
          city: name,
        });
        console.info('Finished fetching data for location:', location);
      } catch (error) {
        if (error instanceof Error && error.message === 'Location not found') {
          setError('Location not found');
        } else if (error instanceof Error) {
          setError('Oops, unexpected error!');
        }

        console.error('Error fetching data for location:', error);
      }
    },
    [getWeatherData]
  );

  const sharedState = useMemo<GlobalContextInterface>(
    () => ({
      state,
      error,
      setError,
      fetchData,
      handleSearch,
    }),
    [state, error, setError, fetchData, handleSearch]
  );

  return (
    <GlobalContext.Provider value={sharedState}>
      {children}
    </GlobalContext.Provider>
  );
};
