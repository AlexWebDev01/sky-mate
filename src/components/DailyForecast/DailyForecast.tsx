import { useGlobalContext } from '@context/global/GlobalContext';

import DailyForecastItem from './components/DailyForecastItem';

import { DailyWeatherData } from '@api/fetchWeatherData/fetchWeatherData.interface';

import './DailyForecast.scss';

interface IProps {
  expanded: boolean;
  onExpand: () => void;
}

const DailyForecast = ({ expanded, onExpand }: IProps) => {
  const { state } = useGlobalContext();
  const { weatherData } = state;

  if (!weatherData) {
    return null;
  }

  return (
    <div
      className={expanded ? 'forecast-expanded' : 'daily-forecast'}
      onClick={onExpand}
    >
      {weatherData.daily.map((item: DailyWeatherData, index: number) => {
        return <DailyForecastItem item={item} key={index} />;
      })}
    </div>
  );
};

export default DailyForecast;
