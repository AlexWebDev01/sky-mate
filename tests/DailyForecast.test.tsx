import '@testing-library/jest-dom';
import {
  fireEvent,
  queryByTestId,
  render,
  screen,
} from '@testing-library/react';
import { GlobalProvider } from '@context/global/GlobalContext';
import DailyForecast from '@components/DailyForecast/DailyForecast';
import { WeatherData } from '@api/fetchWeatherData/fetchWeatherData.interface';
import { WeatherConditions } from '@shared/constants/clothesAlgorithm/clothesAlgorithm.interface';
import { useState } from 'react';

const mockWeatherData: WeatherData = {
  current: {
    clouds: 75,
    dew_point: 10,
    dt: 1631990400,
    feels_like: 12,
    humidity: 50,
    pressure: 1000,
    sunrise: 1631990400,
    sunset: 1632034400,
    temp: 15,
    uvi: 5,
    visibility: 10000,
    weather: [
      {
        main: WeatherConditions.clouds,
        description: 'overcast clouds',
        icon: '04d',
        id: 804,
      },
    ],
    wind_deg: 180,
    wind_speed: 2.5,
  },
  daily: [
    {
      dt: 1631990400,
      sunrise: 1631990400,
      sunset: 1632034400,
      moonrise: 1631990400,
      moonset: 1632034400,
      moon_phase: 0.5,
      temp: { day: 15, min: 10, max: 20, night: 10, eve: 12, morn: 10 },
      feels_like: { day: 12, night: 10, eve: 12, morn: 10 },
      pressure: 1000,
      humidity: 50,
      dew_point: 10,
      wind_speed: 2.5,
      wind_deg: 180,
      wind_gust: 3.5,
      clouds: 75,
      uvi: 5,
      pop: 0.5,
      rain: 0.5,
      summary: 'overcast clouds',
      weather: [
        {
          main: WeatherConditions.sun,
          description: 'clear sky',
          icon: '01d',
          id: 800,
        },
      ],
    },
  ],
  hourly: [
    {
      dt: 1631990400,
      temp: 15,
      feels_like: 12,
      pressure: 1000,
      humidity: 50,
      dew_point: 10,
      uvi: 5,
      clouds: 75,
      visibility: 10000,
      wind_speed: 2.5,
      wind_deg: 180,
      wind_gust: 3.5,
      pop: 0.5,
      weather: [
        {
          main: WeatherConditions.clouds,
          description: 'overcast clouds',
          icon: '04d',
          id: 804,
        },
      ],
    },
  ],
  lat: 56.946,
  lon: 24.1059,
  timezone: 'Europe/Riga',
  timezone_offset: 10800,
};

const DailyForecastParent = () => {
  const [expanded, setExpanded] = useState(false);

  const handleExpand = () => {
    setExpanded(!expanded);
  };

  return <DailyForecast expanded={expanded} onExpand={handleExpand} />;
};

jest.mock('@context/global/GlobalContext', () => ({
  useGlobalContext: jest.fn(),
  GlobalProvider: ({ children }: { children: React.ReactNode }) => (
    <div>{children}</div>
  ),
}));

describe('DailyForecast', () => {
  beforeEach(() => {
    const { useGlobalContext } = jest.requireMock(
      '@context/global/GlobalContext'
    );
    useGlobalContext.mockImplementation(() => ({
      state: {
        weatherData: mockWeatherData,
        isLoading: false,
        coordinates: { lat: 56.946, lon: 24.1059 },
        location: 'Riga',
        pageStyle: 'clouds',
        clothesAdvice: 'Wear a jacket',
      },
      error: '',
      setError: jest.fn(),
      fetchData: jest.fn(),
      handleSearch: jest.fn(),
    }));
  });

  test('renders without crashing', () => {
    render(
      <GlobalProvider>
        <DailyForecast expanded={false} onExpand={jest.fn()} />
      </GlobalProvider>
    );
    expect(screen.getByTestId('daily-forecast')).toBeInTheDocument();
  });

  test('does not render if weatherData is null', () => {
    const { useGlobalContext } = jest.requireMock(
      '@context/global/GlobalContext'
    );
    useGlobalContext.mockImplementation(() => ({
      state: {
        weatherData: null,
        isLoading: false,
        coordinates: { lat: 56.946, lon: 24.1059 },
        location: 'Riga',
        pageStyle: 'clouds',
        clothesAdvice: 'Wear a jacket',
      },
      error: '',
      setError: jest.fn(),
      fetchData: jest.fn(),
      handleSearch: jest.fn(),
    }));

    const { container } = render(
      <GlobalProvider>
        <DailyForecast expanded={false} onExpand={jest.fn()} />
      </GlobalProvider>
    );
    expect(queryByTestId(container, 'daily-forecast')).toBeNull();
  });

  test('toggles expansion on click', () => {
    render(<DailyForecastParent />);

    const dailyForecast = screen.getByTestId('daily-forecast');
    expect(dailyForecast).toHaveClass('daily-forecast');

    fireEvent.click(dailyForecast);
    expect(dailyForecast).toHaveClass('forecast-expanded');

    fireEvent.click(dailyForecast);
    expect(dailyForecast).toHaveClass('daily-forecast');
  });

  test('matches snapshot', () => {
    const { asFragment } = render(
      <DailyForecast expanded={false} onExpand={jest.fn()} />
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
