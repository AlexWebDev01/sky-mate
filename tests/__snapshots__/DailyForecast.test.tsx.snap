// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DailyForecast matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="daily-forecast"
    data-testid="daily-forecast"
  >
    <div
      class="daily-weather"
      data-testid="daily-weather"
    >
      <div
        class="day"
      >
        Sat
      </div>
      <div
        class="sun date"
      >
        18 sept
      </div>
      <div
        class="day-temperature"
      >
        15°
      </div>
      <div
        class="night-temperature"
      >
        10°
      </div>
      <div
        class="weather"
      >
        sun
      </div>
    </div>
  </div>
</DocumentFragment>
`;
